Build: Use globalThis over process.env and enable TS lib checking

sirreal opened the pull request:
## What?

There have been some problems with types in some packages. This resulted in build errors and in producing types that were not valid. This was observed in several other PRs recently.

This PR does two things:
- Remove all `skipLibCheck` from tsconfig.
- Update usage of custom `process.env.NAME` usage in the application to `globalThis.NAME`. This has the benefit of being safer in third party code and valid in the type system.

There are a few problems with using `process.env`, although it's common to use some names for historic reasons - primarily `process.env.NODE_ENV`.

- It only exists in Node.js. It does not exist in the browser. This means that additional checks must be added in application code to make sure that it doesn't error. Using `process.env` in general is error prone.
- [`process.env` in Node.js is unusual. It's a key-value structure where _the values are always strings_.](https://nodejs.org/docs/latest-v20.x/api/process.html#processenv) Gutenberg used non-string values which was wrong in environments where process.env existed, but also did compile-time replacement sometimes. This makes everything confusing.
- The inconsistent process.env values caused problems in the type system. They conflicted with the Node.js types that _correctly assert that process.env values must be `strings`_.

Instead of using the problematic `process.env` "object" for these values, we change to use [`globalThis`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis). This is the global object that works natively in both browsers and Node.js. It is an object and can be used as such, so assigning arbitrary properties is straightforward.

One big benefit is that `globalThis` is much easier to use, `globalThis.nonExistentProperty` is `undefined` instead of a `ReferenceError` like when trying to access `process.env.nonExistentProperty` in the browser.

This eliminates the type issues around using `process.env` and is generally more coherent. The build tooling and code using WordPress and Gutenberg variables around `process.env` has been updated in this PR. Documentation has also been updated where I found it.

Some eslint rules have also been added, changed, or removed:

- `@wordpress/wp-global-usage` replaces `@wordpress/is-gutenberg-plugin`. It applies the same logic to all the globals that Gutenberg uses - namely that they should be used in simple conditional tests to help with dead code elimination in compiled packages.
- `@wordpress/wp-process-env` is a new rule that makes the usage of `process.env` WordPress/Gutenberg variables a violation. It includes a fixer, so e.b. `process.env.SCRIPT_DEBUG` becomes `globalThis.SCRIPT_DEBUG`.
- `@wordpress/gutenberg-phase` rule is removed - this variable has not been used for some time.


Related changes:
- [x] Two packages need to be updated. (#61572, #61503)
- [x] One packages needs to be patched. (#61570)



## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

sirreal committed:
No more skipLibCheck

sirreal committed:
Fix the process problem

sirreal committed:
Go bonkers on the global vars

youknowriad commented on packages/private-apis/src/implementation.js:
```diff
@@ -66,7 +66,7 @@ let allowReRegistration;
 // environment variable is not explicitly defined.
 try {
-	allowReRegistration = process.env.IS_WORDPRESS_CORE ? false : true;
+	allowReRegistration = globalThis.IS_WORDPRESS_CORE ? false : true;
```
These `process.env`  variables are supposed to be tree shakable and if I'm not wrong web pack replaces `process.env.VARIABLE` with the right value in order to be able to tree-shake the build, which mean I'm not sure globalThis works for these variables.

sirreal replied:
There isn't anything special about `process.env` that get special optimizations. Different compilers to provide some special defaults, but `process.env.WHATEVER` is likely more dangerous to use than `globalThis.WHATEVER`. Browsers don't have `process`, so using `process.env.SOMETHING` without an appropriately configured compiler has a strong chance of causing errors. https://github.com/WordPress/gutenberg/pull/16679 was a good example of this.

---

This PR is incomplete, but part of this involved fixing type issues with process. We claim some values are booleans, but this is [actually impossible in node](https://nodejs.org/docs/latest-v20.x/api/process.html#processenv). `process.env.ANYTHING` is type `undefined | string`:

>Assigning a property on process.env will implicitly convert the value to a string. This behavior is deprecated. Future versions of Node.js may throw an error when the value is not a string, number, or boolean.

It doesn't make much sense in the browser because the browser… doesn't have `process`. The thing that saves us is that this is handled by replacement in webpack, but it doesn't make a lot of sense to continue using webpack replacements in this way when there are better alternatives.

---

`globalThis` should be safer to use in published packages. Right now packages include some globals that are not very safe, like `SCRIPT_DEBUG`. Props to @gziolo for the `globalThis` idea.

In published packages, this isn't much different because no replacement is happening right now. In fact, this approach should be safer because `globalThis` _should be_ an object in all contexts I'm aware of and it will save needing to do checks like `typeof VAR !== 'undefined' && VAR` everywhere.

These things _do_ end up in packages published on npm, [for example the block-editor package](https://unpkg.com/browse/@wordpress/block-editor@12.25.0/build-module/components/block-controls/slot.js).

---

Regarding code elimination, I'll speak about webpack because it's what I'm most familiar with. I believe other systems work similarly.

_ASIDE_ I just noticed we have a [babel plugin](https://github.com/WordPress/gutenberg/blob/trunk/packages/warning/babel-plugin.js) that replaces the `warning()` package calls in the compiled packages so that they actually contain: `if ( typeof SCRIPT_DEBUG !== 'undefined' && SCRIPT_DEBUG === true ) {/*…*/}`. That also needs to be updated in this PR, so even where `SCRIPT_DEBUG` does not appear in source, it will appear in compiled packages for npm.

The systems must be configured to do replacement. Webpack does this with its `DefinePlugin`. That's why this PR includes these changed lines:

https://github.com/WordPress/gutenberg/blob/40c0fd7528d4d576d8e9e5ea96640f1eb4c2f4a1/tools/webpack/shared.js#L65-L74

As webpack compiles the code, it does replacement `DefinePlugin`. So when webpack encounters `globalThis.SCRIPT_DEBUG` it replaces it completely with `true`:

```js
if ( globalThis.SCRIPT_DEBUG ) { doThing() }
// becomes
if ( true ) { doThing() }
// or maybe
if ( false ) { doThing() }
```

The above example from the block editor package looks like this:
```js
// input
globalThis.SCRIPT_DEBUG === true ? warning(`Unknown BlockControls group "${group}" provided.`) : void 0;
// compiled (with webpack names replaced)
true ? warning(`Unknown BlockControls group "${group}" provided.`) : 0;
```

Depending on the compilation configuration, this may be the final result and what's included in the package. This is the first part of the process.

Some builds will optimize this code, which is where code is actually eliminated. After optimization, we wind up with:

```js
// true case
warning(`Unknown BlockControls group "${group}" provided.`);
// false case
// …condition and block are completely eliminated…
```

gziolo replied:
I can confirm that tree-shaking works as expected with the changes applied.

sirreal committed:
Update warning package

sirreal committed:
Add globalThis.SCRIPT_DEBUG replacement to scripts

sirreal committed:
Better replacement

Webpack replacements are source string replacements

sirreal committed:
Update tests

gziolo commented on packages/interactivity/src/directives.js:
```diff
@@ -242,11 +242,7 @@ export default () => {
-						SCRIPT_DEBUG === true &&
-						! isPlainObject( value )
-					) {
+					if ( globalThis.SCRIPT_DEBUG && ! isPlainObject( value ) ) {
```
I'm really surprised that `globalThis` isn't throwing any errors but at the same time it simplifies everything:

<img width="404" alt="Screenshot 2024-05-10 at 09 20 29" src="https://github.com/WordPress/gutenberg/assets/699132/157d4463-917d-431e-9d1c-25026a2a4cbe">

gziolo commented on patches/react-autosize-textarea+7.1.0.patch:
```diff
@@ -0,0 +1,13 @@
+diff --git a/node_modules/react-autosize-textarea/lib/TextareaAutosize.d.ts b/node_modules/react-autosize-textarea/lib/TextareaAutosize.d.ts
```
`react-autosize-textarea` is probably no longer maintained because the GitHub repo is archived:
https://github.com/buildo/react-autosize-textarea

We should consider forking it if we need to apply fixes, as it might be easier to change the source file:

https://github.com/buildo/react-autosize-textarea/blob/master/src/TextareaAutosize.tsx

sirreal replied:
This patch is actually pretty simple although it doesn't look like it. It removes two keys that were removed from the React types that were causing issues.

I agree, forking it is a better long term solution so we actually have a maintained version.

gziolo replied:
The patch is simple, but applying further changes might be surprising because of using the patch command.

gziolo replied:
There are other issues with this lib. I found the related issue: https://github.com/WordPress/gutenberg/issues/48009. It's also where I learned about the repository being archived. A related comment from @t-hamano about exploring the possibility to replace the library: https://github.com/WordPress/gutenberg/issues/48009#issuecomment-1735720752.

t-hamano replied:
In the past, we have attempted to replace this package with another package. However, it was not possible to completely solve the problem. The PR fixing the bug that exists on the library side has also not been merged yet, so that attempt has been closed.

I was also hoping that [CSS `field-sizing`](https://developer.chrome.com/docs/css-ui/css-field-sizing) might be able to completely replace this library, but it hasn't been implemented in all major browsers yet.

We might be better off simulating the behavior of this library or forking this library, as mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/48215#issuecomment-1827645994).

sirreal replied:
I plan to pull this patch out to its own PR as a quick fix to an immediate problem. Other than that I'm not very familiar with this library or what it's used for.

I've created an issue to track work on a better solution: https://github.com/WordPress/gutenberg/issues/61568

sirreal replied:
Extracting here: https://github.com/WordPress/gutenberg/pull/61570

gziolo commented on tools/webpack/shared.js:
```diff
@@ -64,13 +64,15 @@ const plugins = [
 		// Inject the `IS_GUTENBERG_PLUGIN` global, used for feature flagging.
-		'process.env.IS_GUTENBERG_PLUGIN':
-			process.env.npm_package_config_IS_GUTENBERG_PLUGIN,
+		'globalThis.IS_GUTENBERG_PLUGIN': JSON.stringify(
```
I like the direction. We will have to remember to update the WordPress Core so tree-shaking continues to work.

youknowriad replied:
Adding the "php backport" label. Although it's not really PHP :)

ellatrix replied:
What is to be backported? Is this file shared with core?

youknowriad replied:
The webpack config change.

sirreal replied:
It would be a very small and simple (but important!) change. I'll open a PR for that soon so it's ready.

sirreal replied:
I've opened https://core.trac.wordpress.org/ticket/61262 and https://github.com/WordPress/wordpress-develop/pull/6596.

gziolo reviewed (commented):
I like the direction. Thank you for taking care of the additional refactoring for the special constants we use to guard some features like `IS_GUTENBERG_PLUGIN`, `IS_WORDPRESS_CORE` and `SCRIPT_DEBUG`. 

One file still needs to get updated:

https://github.com/WordPress/gutenberg/blob/trunk/test/unit/config/gutenberg-env.js

If I follow it correctly, it sets the default values for unit tests, now it should be set on `globalThis.*`.

sirreal commented:
> I like the direction. Thank you for taking care of the additional refactoring for the special constants we use to guard some features like `IS_GUTENBERG_PLUGIN`, `IS_WORDPRESS_CORE` and `SCRIPT_DEBUG`.

Thanks! I agree, I think this will be easier to work with 👍 
 
> One file still needs to get updated:

Good spot! Maybe we should also switch the global `global` to be `globalThis`.

sirreal commented:
There are several eslint rules around how to use `process.env.[OUR_VARIABLES]`. They probably need to be updated to _ban_ those variables (maybe autofix them to the `globalThis…` versions) and warn in the same cases (usage outside of conditionals):

https://github.com/WordPress/gutenberg/blob/5915f9ad0ebc69d77ee99769647a3398920d80ba/packages/eslint-plugin/rules/is-gutenberg-plugin.js#L80-L84

https://github.com/WordPress/gutenberg/blob/5915f9ad0ebc69d77ee99769647a3398920d80ba/packages/eslint-plugin/rules/gutenberg-phase.js#L162-L172

(Maybe GUTENBERG_PHASE needs to be removed or updated to simply ban the rule?)

sirreal committed:
Some env improvements

sirreal committed:
Add no-wp-process-env rule

sirreal committed:
Implement fixer

sirreal committed:
Update lint and add noGutenbergPhase rule

sirreal committed:
Remove gutenberg-phase eslint

sirreal committed:
Update tests

sirreal committed:
rework is-gb-plugin

sirreal committed:
Replace process.env.IS_GUTENBERG_PLUGIN usage

sirreal committed:
docs

sirreal committed:
Add eslint-plugin changelog

sirreal committed:
Update eslint plugin README

sirreal committed:
Update eslint-ignore

sirreal committed:
Fix property name

sirreal committed:
Fix docs

sirreal committed:
Add basic fixing

sirreal committed:
Use better fix

sirreal committed:
fixin the rules

sirreal committed:
Fix warning package lints

sirreal committed:
Fix test file

sirreal committed:
Update docs

sirreal committed:
fix IS_WORDPRESS_CORE

sirreal committed:
eslint plugin

sirreal committed:
lint disable

sirreal commented:
This should have breaking change changelog for all the updated packages- maybe for all packages- mentioning the change to `globalThis`.

sirreal commented on packages/eslint-plugin/CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
+### Breaking Changes
+
+-   `@wordpress/is-gutenberg-plugin` rule has been replaced by `@wordpress/wp-global-usage` ([#61486](https://github.com/WordPress/gutenberg/pull/61486)).
+-   `@wordpress/wp-process-env` rule has been added and included in the recommended configurations ([#61486](https://github.com/WordPress/gutenberg/pull/61486)).
```
`@wordpress/gutenberg-phase` rule removed. That doesn't exist and hasn't been used for some time.

gziolo replied:
Yes, it stopped being a thing after WordPress Phase 2 was concluded - sometime between WP 5.9 and 6.3 when site editor was going out of beta.

sirreal committed:
Add gutenberg-phase changelog

sirreal commented on packages/eslint-plugin/rules/no-wp-process-env.js:
```diff
@@ -0,0 +1,93 @@
+		'GUTENBERG_PHASE',
+		'IS_GUTENBERG_PLUGIN',
+		'IS_WORDPRESS_CORE',
+		'SCRIPT_DEBUG',
```
Technically I think `SCRIPT_DEBUG` was never part of `process.env`, but I don't think its harmful to include it.

gziolo replied:
Yes, when `SCRIPT_DEBUG` was introduced we knew that `process.env` is no longer a way forward.

sirreal commented on docs/how-to-guides/feature-flags.md:
```diff
@@ -71,7 +74,7 @@ if ( undefined ) { // Wepack has replaced `process.env.IS_GUTENBERG_PLUGIN` with
 ### Dead code elimination
 
-For production builds, webpack ['minifies'](https://en.wikipedia.org/wiki/Minification_(programming)) the code, removing as much unnecessary JavaScript as it can. 
+For production builds, webpack ['minifies'](<https://en.wikipedia.org/wiki/Minification_(programming)>) the code, removing as much unnecessary JavaScript as it can.
```
Bad angle brackets in link here.

sirreal replied:
I _think_ this is actually prettier fixing the URL? At any rate, parens in link URLs in markdown is confusing and it's a good idea to URL encode them.

Moving to https://github.com/WordPress/gutenberg/pull/61675

sirreal commented on packages/components/src/custom-gradient-picker/index.tsx:
```diff
@@ -79,7 +79,7 @@ const GradientTypePicker = ( {
 					? undefined
 					: HORIZONTAL_GRADIENT_ORIENTATION,
 				type: 'linear-gradient',
-			} as gradientParser.LinearGradientNode )
```
This could be its own PR.

sirreal replied:
https://github.com/WordPress/gutenberg/pull/61679

gziolo commented on docs/how-to-guides/feature-flags.md:
````diff
@@ -32,37 +33,39 @@ If you're attempting to import and call a plugin-only feature, be sure to wrap t
 ```js
-if ( undefined ) { // Wepack has replaced `process.env.IS_GUTENBERG_PLUGIN` with `undefined`
+if ( undefined ) {
+	// Wepack has replaced `globalThis.IS_GUTENBERG_PLUGIN` with `undefined`
````
Nit 😀

```suggestion
	// Webpack has replaced `globalThis.IS_GUTENBERG_PLUGIN` with `undefined`
```

sirreal committed:
Fix typo

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

sirreal commented:
Some conversation happened in Core Slack around the proposed approach. I'll link here for posterity:

https://wordpress.slack.com/archives/C02QB2JS7/p1715758629398369

sirreal committed:
Fix url encoding of parens

sirreal committed:
Revert unrelated markdown changes

sirreal commented:
> Is there any make/core post or dev note that should be done as a result of this PR. Who would be interested to learn about this and how should we document it.

We could post on make/core about it, I can draft something if you think that's the best place.

I was also wondering about making a breaking change for packages. Very few packages are actually affected by this change. In practice, these variables don't seem to be used that frequently. But which package? Potentially packages will be affected in the future and I don't expect folks to add a breaking change when they use one of these variables 😕 a

I was thinking about making a breaking change for all the packages that are intended for client side use and linking to some documentation.

Thoughts?

youknowriad commented:
Yeah, marking as a breaking change seems good to me.
For the make/core post I'm not entirely sure to be honest, I'm raising it for discussion. Who needs to know about this. Package consumers seems like an obvious thing (thus the breaking changes) but other than that it's unclear.

sirreal commented:
Do you think all the client packages should get a breaking change or only the ones immediately impacted?

> Who needs to know about this?

Folks that have their own custom bundlers set up that are replacing the forms that are changed in this PR. They'd need to update replacements. So if they have `webpack.DefinePlugin( { 'process.env.IS_GUTENBERG_PLUGIN': 'false' } )` it should become `webpack.DefinePlugin( { 'globalThis.IS_GUTENBERG_PLUGIN': 'false' } )`.

That said, one of the goals here is to ensure that if folks don't do that, everything works.

Folks using wp-scripts should be fine.

youknowriad commented:
> Do you think all the client packages should get a breaking change or only the ones immediately impacted?

I guess the ones that are immediately impacted and the packages that depend on them. I don't think that's equivalent to all packages.

sirreal commented on packages/scripts/config/webpack.config.js:
```diff
@@ -307,7 +307,8 @@ const scriptConfig = {
 			// Inject the `SCRIPT_DEBUG` global, used for development features flagging.
-			SCRIPT_DEBUG: ! isProduction,
+			'globalThis.SCRIPT_DEBUG': JSON.stringify( ! isProduction ),
+			SCRIPT_DEBUG: JSON.stringify( ! isProduction ),
```
I'm surprised we don't define `IS_GUTENBERG_PLUGIN` or `IS_WORDPRESS_CORE` here. Third parties that depend on packages probably want those.

This may be good for a follow-up.

gziolo replied:
Why would they need it? WordPress Core exposes all WP packages as entry points and they are already processed with these constants replaced. I don't think that plugins or theme authors would ever want to use these constants as they aren't changing the code of the Gutenberg plugin or WordPress Core.

sirreal replied:
That's a good point, but it would make a difference for folks using wp-script to bundle things outside WordPress. That's probably very uncommon.

gziolo replied:
We can add it if that is the case. Both probably should be `false` in that case.

gziolo commented:
What type of breaking changes do you expect after refactoring these three constants to be set on `globalThis`? All the code will continue to work. Are you worried that tree-shaking won't work out of the box anymore? I don't mind enforcing a major version bump to account for that. We could also sync that with enforcing some other breaking changes like increasing the minimum version of Node to 20, which is the lowest supported version by Gutenberg, WP core, and also the lowest active LTS version. That way we could draft some sort of dev note / announcement on Make Core explaining all of that in one go.

sirreal commented:
> What type of breaking changes do you expect after refactoring these three constants to be set on `globalThis`? … Are you worried that tree-shaking won't work out of the box anymore?

Exactly this. It's going to be a very small difference, but if I had webpack set up to replace things correctly and then a library switched them without informing me… I'd be annoyed.

> All the code will continue to work.

Definitely, that's what I find most compelling about this change is that on the whole things should be less fragile.

gziolo commented on docs/how-to-guides/feature-flags.md:
```diff
@@ -99,6 +102,6 @@ In this case, the minification process will remove the entire `if` statement inc
 ## Frequently asked questions
 
-### Why shouldn't I assign the result of an expression involving `IS_GUTENBERG_PLUGIN` to a variable, e.g. `const isMyFeatureActive = process.env.IS_GUTENBERG_PLUGIN === 2`?
+### Why shouldn't I assign the result of an expression involving `IS_GUTENBERG_PLUGIN` to a variable, e.g. `const isMyFeatureActive = globalThis.IS_GUTENBERG_PLUGIN === 2`?
```
This needs to be updated, as it no longer fits. The original phrasing was related to the old constant that was a number. Now it's a boolean.

sirreal replied:
```suggestion
### Why shouldn't I assign the result of an expression involving `IS_GUTENBERG_PLUGIN` to a variable, e.g. `const isMyFeatureActive = ! Object.is( undefined, globalThis.IS_GUTENBERG_PLUGIN )`?
```

gziolo replied:
Actually, it probably doesn't matter. I see what this section tries to explain. Do not use it with complex expressions. Now, I'm wondering if tree-shaking would work properly with `const isActive = globalThis.IS_GUTENBERG_PLUGIN; if ( isActive ) { ... }`.

sirreal committed:
Update stale code example

gziolo commented on test/unit/config/gutenberg-env.js:
```diff
@@ -1,32 +1,13 @@
+ * WordPress core.
+ */
+// eslint-disable-next-line @wordpress/wp-global-usage
+globalThis.IS_WORDPRESS_CORE = true;
```
Should we use a similar syntax here as with `IS_GUTENBERG_PLUGIN`? I see that inside the webpack shared config both constants can be sourced from the npm config.

talldan replied:
I think it has to be hard-code to `true` here to override the default for the project (which is falsey).

If a private API is re-registered, the tests should throw an error.

Though I wonder if it'd be better for gutenberg to set this to `true` anyway. At the moment it seems that re-registration is allowed for the plugin build?

I guess it'd be a breaking change now, so not much can be done.

talldan replied:
> At the moment it seems that re-registration is allowed for the plugin build?

Oh, maybe so that the plugin can re-define the modules that core has already defined.

gziolo reviewed (commented):
Code-wise, everything looks great. All the tests pass. I can confirm that code behind the feature flags is getting properly handled.


`globalThis.IS_GUTENBERG_PLUGIN` correctly replaced with `true` in the development mode:

<img width="1394" alt="Screenshot 2024-05-15 at 18 20 10" src="https://github.com/WordPress/gutenberg/assets/699132/0720ecfb-2391-4b04-b3d1-37d8a368ff09">

`globalThis.SCRIPT_DEBUG` correctly replaced with `true` in the development mode:

<img width="1440" alt="Screenshot 2024-05-15 at 18 20 52" src="https://github.com/WordPress/gutenberg/assets/699132/08335344-7d9c-4b82-8152-73850a99d746">

`globalThis.IS_WORDPRESS_CORE` correctly replaced with `false` in the development mode:

<img width="1116" alt="Screenshot 2024-05-15 at 18 21 52" src="https://github.com/WordPress/gutenberg/assets/699132/c3912b75-383a-4676-ba55-ce537b27382d">

`globalThis.SCRIPT_DEBUG` correctly replaced with `true` in the production mode, code fully eliminated from the minified version:

<img width="1393" alt="Screenshot 2024-05-15 at 18 23 32" src="https://github.com/WordPress/gutenberg/assets/699132/9abeecb1-eec2-4e85-a765-cee8a5d88a60">


---

The only remaining question is whether we want to proceed with `globalThis` as discussed on WordPress Slack (link requires registration at https://make.wordpress.org/chat/):

https://wordpress.slack.com/archives/C02QB2JS7/p1715758629398369

We should perform a research how project handle similar situation these days. We covered alternatives like continuing using `process.env.*` or start using `import.meta.*`. However, both of them have issues that `globalThis.*` resolves. If the bundler doesn't replace these contents they will break in certain scenarios. For `process.env.*` that would happen in the browser. For `import.meta`, that would happen outside of ES Module.

gziolo commented:
I was researching options for how different bundlers replace values at the bundle time, and this [section for Rolloup](https://medium.com/stackanatomy/the-ultimate-guide-to-getting-started-with-the-rollup-js-javascript-bundler-2ebec9398656#4095) is very similar to the approach proposed:

```js
// rollup.config.js
import replace from '[@rollup/plugin-replace](http://twitter.com/rollup/plugin-replace)';
const devMode = (process.env.NODE_ENV === 'development');
console.log(`${ devMode ? 'development' : 'production' } mode bundle`);
export default [
  {
    input: './src/main.js',
    plugins: [
      replace({
        values: {
          __HELLO__: 'Hi there',
          __GOODBYE__: 'Bye'
        }
      })
    ],
    output: {
      file: './build/bundle.js',
      format: 'es',
      sourcemap: devMode ? 'inline' : false
    }
  }
];
```

```js
export function hello() {
  console.log('__HELLO__ from a.js');
}
export function goodbye() {
  console.log('__GOODBYE__ from a.js');
}
```

sirreal committed:
Apply autofix

sirreal committed:
Fix the automatic fixes

sirreal committed:
Cleanup unused ts-ignores

sirreal committed:
Add breaking changes to affected packages

sirreal committed:
Autoformat modified changelogs

sirreal commented:
I've rebased this and added breaking change changelogs to all the affected packages.

sirreal commented:
I received some positive feedback about this change in Slack: https://wordpress.slack.com/archives/C02QB2JS7/p1716288162649129?thread_ts=1715758629.398369&cid=C02QB2JS7

No blockers were mentioned.

sirreal commented:
After receiving positive feedback in Slack and no opposition anywhere, I plan to land this soon.

sirreal commented:
I added and removed the dev note label. Dev notes are largely for WordPress developers and these changes should not affect them at all. They'll be using the included Scripts from WordPress that already have the optimizations applied.

This change already includes breaking change changelogs for folks consuming the packages on npm, that's probably the best place for it.

Mamaduka commented:
@sirreal, I see the following error in the editor after this got merged. Is there an additional step to get it fixed?

## Screenshot
![CleanShot 2024-05-23 at 20 01 09](https://github.com/WordPress/gutenberg/assets/240569/9d67440a-69f6-48d0-8286-01e4289368cf)

sirreal commented:
@Mamaduka I think that's probably a stale process or something. Try running `npm ci` and restarting the linter process or your editor. I'm not able to reproduce any issue with the new lint rules in the repo.

Mamaduka commented:
Thanks, @sirreal! The `npm ci` did the trick.
